tempesta_fw/http_stream.c

```diff
@@ -0,0 +1,408 @@
+			stream->state = HTTP2_STREAM_LOC_HALF_CLOSED;
+		break;
+
+	case HTTP2_STREAM_CONT_HC_CLOSED:
+		if (type != HTTP2_CONTINUATION) {
+			*err = HTTP2_ECODE_PROTO;
+			return STREAM_FSM_RES_TERM_CONN;
+		}
+		/*
+		 * If END_HEADERS flag arrived in this state, this means that
+		 * END_STREAM flag had been already received earlier - in
+		 * half-closed (local) state, and now we must close the stream,
+		 * i.e. move it to final 'closed' state.
+		 */
+		if (flags & HTTP2_F_END_HEADERS)
+			stream->state = HTTP2_STREAM_CLOSED;
+		break;
+
+	case HTTP2_STREAM_LOC_HALF_CLOSED:
+		/*
+		 * According to section 5.1 of RFC 7540 any frame type can be
+		 * received and will be valid in 'half-closed (local)' state.
+		 */
+		if (type == HTTP2_HEADERS)
+		{
+			if (flags & (HTTP2_F_END_STREAM | HTTP2_F_END_HEADERS)) {
+				stream->state = HTTP2_STREAM_CLOSED;
+			}
+			else if (flags & HTTP2_F_END_STREAM) {
+				stream->state = HTTP2_STREAM_CONT_HC_CLOSED;
+			}
+			else if (!(flags & HTTP2_F_END_HEADERS)) {
+				stream->state = HTTP2_STREAM_CONT_HC;
+			}
+		}
+		else if ((type == HTTP2_DATA && (flags & HTTP2_F_END_STREAM))
+			|| type == HTTP2_RST_STREAM)
+		{
+			stream->state = HTTP2_STREAM_CLOSED;
+		}
+		else if (type == HTTP2_CONTINUATION)
+		{
+			*err = HTTP2_ECODE_PROTO;
+			return STREAM_FSM_RES_TERM_CONN;
+		}
+		break;
+
+	case HTTP2_STREAM_CLOSED:
+		/*
+		 * In moment when the final 'closed' state is achieved, stream
+		 * actually must be removed from stream's storage (and from
+		 * memory), thus the execution flow must not reach this point.
+		 */
+	default:
+		BUG();
+	}
+
+	T_DBG3("exit %s: stream->state=%d\n", __func__, stream->state);
+
+	return STREAM_FSM_RES_OK;
+}
+
+bool
+tfw_http2_stream_is_closed(TfwStream *stream)
+{
+	return stream->state == HTTP2_STREAM_CLOSED;
+}
+
+static inline void
+tfw_http2_init_stream(TfwStream *stream, unsigned int id, unsigned short weight)
+{
+	RB_CLEAR_NODE(&stream->node);
+	stream->id = id;
+	stream->state = HTTP2_STREAM_OPENED;
+	stream->weight = weight ? weight : HTTP2_DEF_WEIGHT;
+}
+
+TfwStream *
+tfw_http2_find_stream(TfwStreamSched *sched, unsigned int id)
+{
+	struct rb_node *node = sched->streams.rb_node;
+
+	while (node) {
+		TfwStream *stream = rb_entry(node, TfwStream, node);
+
+		if (id < stream->id)
+			node = node->rb_left;
+		else if (id > stream->id)
+			node = node->rb_right;
+		else
+			return stream;
+	}
+
+	return NULL;
+}
+
+TfwStream *
+tfw_http2_add_stream(TfwStreamSched *sched, unsigned int id,
+		     unsigned short weight)
+{
+	TfwStream *new_stream;
+	struct rb_node **new = &sched->streams.rb_node;
+	struct rb_node *parent = NULL;
+
+	while (*new) {
+		TfwStream *stream = rb_entry(*new, TfwStream, node);
+
+		parent = *new;
+		if (id < stream->id) {
+			new = &parent->rb_left;
+		} else if (id > stream->id) {
+			new = &parent->rb_right;
+		} else {
+			WARN_ON_ONCE(1);
+			return NULL;
+		}
+	}
+
+	new_stream = kzalloc(sizeof(TfwStream), GFP_ATOMIC);
```

Comment:
The same allocation problem: basically we usually don't use pattern `kzalloc(sizeof(....), ...)` in any more or less performance crucial code - if we have fixed size allocations, then we use some kind of SLAB allocator.

Reply:
Yeah - missed that. Thanks!

Reply:
Corrected.